sm/download/src/download.rs

```diff
@@ -142,6 +191,104 @@ mod tests {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn downloader_download_with_content_length() -> anyhow::Result<()> {
+        let tmpstats = statvfs::statvfs("/tmp")?;
+        let usable_disk_space = tmpstats.blocks_free() * tmpstats.block_size();
+        let _mock1 = mock("GET", "/some_file.txt")
+            .with_header("content-length", &(usable_disk_space.to_string()))
+            .create();
+
+        let name = "test_download_with_length";
+        let version = Some("test1".to_string());
+        let target_dir_path = TempDir::new()?;
+
+        let mut target_url = mockito::server_url();
+        target_url.push_str("/some_file.txt");
+
+        let url = DownloadInfo::new(&target_url);
+
+        let downloader = Downloader::new(&name, &version, target_dir_path.path());
+        match downloader.download(&url).await {
+            Err(DownloadError::InsufficientSpace) => return Ok(()),
+            _ => bail!("failed"),
+        }
+    }
+
+    #[tokio::test]
+    async fn downloader_download_with_reasonable_content_length() -> anyhow::Result<()> {
+        let file = create_file_with_size(10 * 1024 * 1024)?;
+
+        let _mock1 = mock("GET", "/some_file.txt")
+            .with_body_from_file(file.into_temp_path())
+            .create();
+
+        let name = "test_download_with_length";
+        let version = Some("test1".to_string());
+        let target_dir_path = TempDir::new()?;
+
+        let mut target_url = mockito::server_url();
+        target_url.push_str("/some_file.txt");
+
+        let url = DownloadInfo::new(&target_url);
+
+        let downloader = Downloader::new(&name, &version, target_dir_path.path());
+        match downloader.download(&url).await {
+            Ok(()) => return Ok(()),
+
+            _ => bail!("failed"),
+        }
+    }
+
+    #[tokio::test]
+    async fn downloader_download_verify_file_content() -> anyhow::Result<()> {
+        let file = create_file_with_size(10)?;
+
+        let _mock1 = mock("GET", "/some_file.txt")
+            .with_body_from_file(file.into_temp_path())
+            .create();
+
+        let name = "test_download_with_length";
+        let version = Some("test1".to_string());
+        let target_dir_path = TempDir::new()?;
+
+        let mut target_url = mockito::server_url();
+        target_url.push_str("/some_file.txt");
+
+        let url = DownloadInfo::new(&target_url);
+
+        let downloader = Downloader::new(&name, &version, target_dir_path.path());
+        downloader.download(&url).await?;
+
+        let log_content = std::fs::read(downloader.filename())?;
+
+        assert_eq!("Some data!".as_bytes(), log_content);
```

Comment:
Ok this test checks the downloaded content, but this should be done for the other tests : they exercise different code paths.